src/treq/client.py

```diff
@@ -141,15 +144,28 @@ def delete(self, url, **kwargs):
         """
         return self.request('DELETE', url, **kwargs)
 
-    def request(self, method, url, **kwargs):
+    def request(self, method, url,
+                headers=None, params=None, data=None, files=None,
+                auth=None, cookies=None, allow_redirects=True,
+                browser_like_redirects=False, unbuffered=False,
+                timeout=None, reactor=None, json=_NO_JSON, **kwargs):
         """
         See :func:`treq.request()`.
         """
+        if kwargs:
+            warnings.warn("".join([
+                "treq.client.HttpClient.request will only accept ",
+                "supported documented arguments. ",
+                "This will raise an Exception in the future. ",
+                "Unsupported arguments: [",
+                ",".join(kwargs.keys()),
+                "]"]),
+                category=DeprecationWarning, stacklevel=2)
```

Comment:
:[  When called directly we need `stacklevel=2` but when called through `treq.request()` etc I think you would want `stacklevel=3`?  I don't know how important this is but we could handle it via a silly stack traversal until we got out of treq.  At least then it would be isolated here and not require passing params etc.  I was curious to practice so I made up this...  I tend to think it's kind of icky and this issue may not be worth addressing but...  the result is nice.  Hopefully the `try`/`finally` would successfully make the worst case be that you get a warning but don't know quite where from.

Py2: https://repl.it/@altendky/GrizzledWatchfulPostgres-3
Py3: https://repl.it/@altendky/GrotesqueAnnualTag-4
```python
import inspect
import warnings

top_level_package = __name__.partition('.')[0]

def warn(*args, **kwargs):
    try:

        stack = inspect.stack()

        if 'stacklevel' not in kwargs:
            for stack_level, frame_info in enumerate(stack, 1):
                module_name = inspect.getmodule(frame_info[0]).__name__
                if not module_name.startswith(top_level_package + '.'):
                    break
            else:
                stack_level = 3
            
            kwargs['stacklevel'] = stack_level
    finally:
        kwargs.setdefault('stacklevel', 3)
        warnings.warn(*args, **kwargs)
```

Reply:
Nice! I think I'd like to see this in twisted (if not python itself, twisted is a good first step).

I added it to `treq._util` with some "improvements", I hope you agree that makes the code more generic :-).

Amongst other things, I noticed that if the code lived somewhere outside of the top_level_package, it'd break the loop on the first frame, also, I think it has a few usability goodies.

Since you caught the difference between `request` and `get`, I went ahead and added a test for all supported methods.